Backend.md

```diff
@@ -0,0 +1,167 @@
+
+Commands live in the `cmd` namespace, found in `commands.hh`. For easy command buffer writing, `command_stream_writer` is provided in the same header.
+
+Commands are almost entirely stateless. The only state is the currently active render pass, marked by `cmd::begin_render_pass` and `cmd::end_render_pass` respectively. Other commands like `cmd::draw`, or `cmd::dispatch` (compute) contain all of the state they require, including their `handle::pipeline_state`.
+
+### Shader Arguments
+
+There are four shader input types, following D3D12 conventions:
+
+1. CBVs - Constant Buffer Views
+
+    A read-only buffer of limited size. HLSL `b` register.
+
+    ```HLSL
+    struct camera_constants {
+        float4x4 view_proj;
+        float3 cam_pos;
+    };
+
+    ConstantBuffer<camera_constants> g_frame_data       : register(b0, space0);
+    ```
+
+2. SRVs - Shader Resource Views
+
+    A read-only texture, or a large, strided buffer. HLSL `t` register.
+
+    ```HLSL
+    StructuredBuffer<float4x4> g_model_matrices         : register(t0, space0);
+    Texture2D g_albedo                                  : register(t1, space0);
+    Texture2D g_normal                                  : register(t2, space0);
+    ```
+
+3. UAVs - Unordered Access Views
+
+    A read-write texture or buffer. HLSL `u` register.
+
+    ```HLSL
+    RWTexture2DArray<float4> g_output                   : register(u0, space0);
+    ```
+
+4. Samplers
+
+    State regarding the way to sample textures. HLSL `s` register.
+
+    ```HLSL
+    SamplerState g_sampler                              : register(s0, space0);
+    ```
+
+Shaders in PRB can be fed with up to 4 "shader arguments". Each shader argument consists of a `handle::shader_view` and a CBV (`handle::resource`). Shader arguments correspond to HLSL spaces. Argument 0 is `space0`, 1 is `space1`, and so forth.
+
+Inputs are not strictly typed, for example, a `Texture2D` can be supplied by simply "viewing" a single array slice of a texture array. Details regarding this process can be inferred from the creation of `handle::shader_view`, and the `shader_view_element` type.
+
+Shader view "shapes", as in the amount of arguments, and the amount of inputs per argument, are supplied when creating a `handle::pipeline_state`. The actual argument values are supplied in commands, like `cmd::draw`.
+
+## Threading
+
+With one exception, the PRB backend is entirely free-threaded and internally synchronized. The synchronization is minimal, and parallel recording of command lists is encouraged, which takes place on thread-local components. The exception is the `Backend::submit` method, which must only be called on one thread at a time.
```

Comment:
is synchronization of `Backend::submit` automatic or must it be guaranteed by the user? It seems to me that automatic sync might be preferable?

Reply:
Yes, we could simply add a mutex to this method. It would be so large that contending on `submit` calls would still be practically a user error, but this would simplify the rules of engagement even more. We might actually be able to warn if submits happen concurrently that way.